executor/aggfuncs/func_value_test.go

```diff
@@ -0,0 +1,107 @@
+		if nth < srcChk.NumRows() {
+			row := srcChk.GetRow(nth - 1)
+			memDeltas[nth-1] = getEvaluatedMemDelta(&row, dataType)
+		}
+		return memDeltas, nil
+	}
+}
+
+func (s *testSuite) TestMemValue(c *C) {
+	tests := []windowMemTest{
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeLonglong, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4IntSize, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeFloat, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4Float32Size, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeDouble, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4Float64Size, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeNewDecimal, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4DecimalSize, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeString, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4StringSize, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeDate, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4TimeSize, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeDuration, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4DurationSize, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncFirstValue, mysql.TypeJSON, 0, 1, 2,
+			aggfuncs.DefPartialResult4FirstValueSize+aggfuncs.DefValue4JSONSize, firstValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncLastValue, mysql.TypeLonglong, 1, 0, 2,
+			aggfuncs.DefPartialResult4LastValueSize+aggfuncs.DefValue4IntSize, lastValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncLastValue, mysql.TypeString, 1, 0, 2,
+			aggfuncs.DefPartialResult4LastValueSize+aggfuncs.DefValue4StringSize, lastValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncLastValue, mysql.TypeJSON, 1, 0, 2,
+			aggfuncs.DefPartialResult4LastValueSize+aggfuncs.DefValue4JSONSize, lastValueEvaluateRowUpdateMemDeltaGens),
+		buildWindowMemTester(ast.WindowFuncNthValue, mysql.TypeLonglong, 2, 0, 3,
+			aggfuncs.DefPartialResult4NthValueSize+aggfuncs.DefValue4IntSize, nthValueEvaluateRowUpdateMemDeltaGens(2)),
+		buildWindowMemTester(ast.WindowFuncNthValue, mysql.TypeLonglong, 5, 0, 3,
+			aggfuncs.DefPartialResult4NthValueSize+aggfuncs.DefValue4IntSize, nthValueEvaluateRowUpdateMemDeltaGens(5)),
+		buildWindowMemTester(ast.WindowFuncNthValue, mysql.TypeJSON, 2, 0, 3,
+			aggfuncs.DefPartialResult4NthValueSize+aggfuncs.DefValue4JSONSize, nthValueEvaluateRowUpdateMemDeltaGens(2)),
+		buildWindowMemTester(ast.WindowFuncNthValue, mysql.TypeString, 5, 0, 3,
```

Comment:
Why the numRows arg of those test is 0?

Reply:
Ah, Sorry... The master branch has a different signature of the same function, but I reused it without carefully checking its consistency when merging.
I'll check it throughly and push them again.

Reply:
I've checked again, the key difference lies on the variable order (a switch between ```numRows``` and ```orderByCols```) of the function ```buildWindowMemTester```.
Thanks very much, again, for your meticulousness and patience.